src/pubsub/mod_pubsub_db_mnesia.erl

```diff
@@ -112,12 +131,136 @@ get_own_nodes_states(JID) ->
                  [], pubsub_state),
     {ok, OwnNodesStates}.
 
+-spec del_state(Nidx :: mod_pubsub:nodeIdx(),
+                JID :: jid:jid()) -> ok.
+del_state(Nidx, JID) ->
+    mnesia:delete({pubsub_state, {jid:to_lower(JID), Nidx}}).
+
+%% ------------------------ Affiliations ------------------------
+
+-spec set_affiliation(Nidx :: mod_pubsub:nodeIdx(),
+                      JID :: jid:jid(),
+                      Affiliation :: mod_pubsub:affiliation()) -> ok.
+set_affiliation(Nidx, JID, Affiliation) ->
+    BareJID = jid:to_bare(JID),
+    {ok, State} = get_state(Nidx, BareJID, write),
+    case {Affiliation, State#pubsub_state.subscriptions} of
+        {none, []} -> del_state(Nidx, BareJID);
+        _ ->  mnesia:write(State#pubsub_state{ affiliation = Affiliation })
+    end.
+
+-spec get_affiliation(Nidx :: mod_pubsub:nodeIdx(),
+                      JID :: jid:jid()) ->
+    {ok, mod_pubsub:affiliation()}.
+get_affiliation(Nidx, JID) ->
+    {ok, State} = get_state(Nidx, jid:to_bare(JID), read),
+    {ok, State#pubsub_state.affiliation}.
+
+%% ------------------------ Subscriptions ------------------------
+
+-spec add_subscription(Nidx :: mod_pubsub:nodeIdx(),
+                       JID :: jid:jid(),
+                       Sub :: mod_pubsub:subscription(),
+                       SubId :: mod_pubsub:subId()) -> ok.
+add_subscription(Nidx, JID, Sub, SubId) ->
+    {ok, State} = get_state(Nidx, JID, write),
+    NSubscriptions = [{Sub, SubId} | State#pubsub_state.subscriptions ],
```

Comment:
Are we sure, that the function caller has made duplicate check?

Reply:
This function is currently called only with brand new IDs and it retains original code behaviour. Unfortunately subscription management in `mod_pubsub` is still a bit cryptic to me (why do we keep a list here?!) so I've decided to leave it as it is.